File: src/question_151/Q169_Majority_Element.java
package question_151;

public class Q169_Majority_Element {
    public int majorityElement(int[] nums) {
        
        int occr = 0;
        int n = 0;
        for(int i : nums){
            if(occr == 0){
                n = i;
                occr = 1;
            } else {
                if(i == n) {
                    occr++;
                } else {
                    occr--;  
                }
            }
        }
        return n;
    }
}

File: js/question_151/Q169_Majority_Element.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var majorityElement = function(nums) {
    // because it appears more than n/2 time, we simply track how many "current" num
    var cur = "",
        times = 0;
    for(var i = 0; i< nums.length; i++){
        if(cur == nums[i]){
            times++;
        }

        else if(times==0) {
            cur = nums[i];
            times = 1;
        } else {
            times--;
        }
    }

    return cur;
};